src/DoctrineListener/DoctrineAuditListener.php

```diff
@@ -0,0 +1,253 @@
+#[AsDoctrineListener(event: Events::postUpdate, priority: 500, connection: 'default')]
+#[AsDoctrineListener(event: Events::preRemove, priority: 500, connection: 'default')]
+#[AsDoctrineListener(event: Events::postRemove, priority: 500, connection: 'default')]
+class DoctrineAuditListener
+{
+    public function __construct(
+        private readonly AuditLogger $auditLogger,
+        private readonly SerializerInterface $serializer,
+        private $removals = [],
+    ) {
+    }
+
+    public function postPersist(PostPersistEventArgs $args): void
+    {
+        $entity = $args->getObject();
+        $entityManager = $args->getObjectManager();
+
+        $this->log($entity, EventAction::CREATE, $entityManager);
+    }
+
+    public function postUpdate(PostUpdateEventArgs $args): void
+    {
+        $entity = $args->getObject();
+        $entityManager = $args->getObjectManager();
+
+        $this->log($entity, EventAction::UPDATE, $entityManager);
+    }
+
+    // We need to store the entity in a temporary array here, because the entity's ID is no longer
+    // available in the postRemove event. We convert it to an array here, so we can retain the ID for
+    // our audit log.
+    public function preRemove(PreRemoveEventArgs $args): void
+    {
+        $entity = $args->getObject();
+        $data = $this->serializer->normalize($entity);
+        $data['trail_fields'] = [];
+
+        // Hide sensitive data
+        $reflectionClass = new ReflectionClass($entity);
+        $properties = $reflectionClass->getProperties();
+        foreach ($properties as $property) {
+            if ($property->getAttributes(AuditTrailSensitiveData::class)) {
+                $property->setAccessible(true);
+                $data[$property->getName()] = '*****';
+            }
+
+            // If the attribute AuditTrailDisplayData::class isn't set, we don't want to log it
+            if ($property->getAttributes(AuditTrailDisplayData::class)) {
+                $data['trail_fields'][] = $property->getName();
+            }
+        }
+
+        $data['trail_id'] = $this->getIdentifierForEntity($entity);
+        $data['entity_id'] = $this->getIdForEntity($entity);
+
+        $this->removals[] = $data;
+    }
+
+    public function postRemove(PostRemoveEventArgs $args): void
+    {
+        $entity = $args->getObject();
+        $entityManager = $args->getObjectManager();
+
+        $this->log($entity, EventAction::DELETE, $entityManager);
+    }
+
+    private function log(
+        object $entity,
+        EventAction $action,
+        EntityManagerInterface $entityManager,
+    ): void {
+        $className = get_class($entity);
+        $entityIdentifier = $this->getIdentifierForEntity($entity);
+        $entityId = $this->getIdForEntity($entity);
+
+        switch ($action) {
+            case EventAction::DELETE:
+                $entityData = array_pop($this->removals);
```

Comment:
Wat gebeurt er bij verwijderen van meerdere entities?

Reply:
Gebeurd sequentieel door de entity manager, dus gaat steeds dezelfde volgorde zijn

Reply:
Was meer in geval als er cascade delete gebeurd, die array_pop leek mij nogal gevaarlijk om de juiste entry te hebben

Reply:
fixed afe614fb2f047adeb3e94c2d82fc8c56be48c7dc